src/storage/src/client/read_object.rs

```diff
@@ -413,11 +416,14 @@ impl ReadObjectResponse {
 
     #[cfg(feature = "unstable-stream")]
     #[cfg_attr(docsrs, doc(cfg(feature = "unstable-stream")))]
-    /// Convert the response to a [futures::Stream].
-    pub fn into_stream(self) -> impl futures::Stream<Item = Result<bytes::Bytes>> {
-        // TODO(#2049): implement checksum checks for streams.
+    /// Convert the response to a [Stream].
+    pub fn into_stream(self) -> impl Stream<Item = Result<bytes::Bytes>> {
         use futures::TryStreamExt;
-        self.inner.bytes_stream().map_err(Error::io)
+        let response_crc32c = self.check_crc32c();
+        Crc32CStream::new(
+            self.inner.bytes_stream().map_err(Error::io),
+            response_crc32c,
+        )
```

Comment:
Or maybe something based on this:

```suggestion
        use futures::StreamExt;
        let stream = self.inner.bytes_stream().map_err(Error::io);
        futures::stream::unfold(Some((stream, 0_u32)), async move |s| {
            if let Some((mut stream, mut crc32c)) = s {
                return match stream.next().await {
                    None =>  if crc32c == 42 { None } else { Some((Err(Error::deser("bad checksum")), None)) }
                    Some(Err(e)) => Some((Err(e), None)),
                    Some(Ok(b)) => { crc32c += 1; Some((Ok(b), Some((stream, crc32c)))) },
                };
            }
            None
        })
```